### PR TITLE
Add a Java-style builder for Driver parameters

### PR DESCRIPTION
**Pull request by exFalso**

### CODE REVIEW

File: test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt [hunk 1]
```diff
@@ -244,32 +244,67 @@ sealed class PortAllocation {
...
-        isDebug: Boolean = false,
-        driverDirectory: Path = Paths.get("build", getTimestampAsDirectoryName()),
-        portAllocation: PortAllocation = PortAllocation.Incremental(10000),
-        debugPortAllocation: PortAllocation = PortAllocation.Incremental(5005),
-        systemProperties: Map<String, String> = emptyMap(),
-        useTestClock: Boolean = false,
-        initialiseSerialization: Boolean = true,
-        networkMapStartStrategy: NetworkMapStartStrategy = NetworkMapStartStrategy.Dedicated(startAutomatically = true),
-        startNodesInProcess: Boolean = false,
+        defaultParameters: DriverParameters = DriverParameters(),
```
Comment (joeldudleyr3) [inline test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt]: Why pass `defaultParameters` as the first argument here? Isn't that param useless?

File: test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt [hunk 2]
```diff
@@ -244,32 +244,67 @@ sealed class PortAllocation {
...
+        val portAllocation: PortAllocation = PortAllocation.Incremental(10000),
+        val debugPortAllocation: PortAllocation = PortAllocation.Incremental(5005),
+        val systemProperties: Map<String, String> = emptyMap(),
+        val useTestClock: Boolean = false,
+        val initialiseSerialization: Boolean = true,
+        val networkMapStartStrategy: NetworkMapStartStrategy = NetworkMapStartStrategy.Dedicated(startAutomatically = true),
+        val startNodesInProcess: Boolean = false
 )
 
+class DriverBuilder(val parameters: DriverParameters) {
```
Comment (joeldudleyr3) [inline test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt]: Why have `parameters` as a constructor argument? Why not just create a `parameters` field?

File: test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt (hunk 1 above)
Comment (exFalso) [reply test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt]: That parameter is used to fill in the other default parameters. This is so we have a single place that defines the default values (in `DriverParameters`)

File: test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt (hunk 2 above)
Comment (exFalso) [reply test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt]: This is required for the builder's setter-chaining (note how each setter returns a *new* builder)

File: test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt [hunk 3]
```diff
@@ -242,33 +243,77 @@ sealed class PortAllocation {
...
+    fun setPortAllocation(portAllocation: PortAllocation) = copy(portAllocation = portAllocation)
+    fun setDebugPortAllocation(debugPortAllocation: PortAllocation) = copy(debugPortAllocation = debugPortAllocation)
+    fun setSystemProperties(systemProperties: Map<String, String>) = copy(systemProperties = systemProperties)
+    fun setUseTestClock(useTestClock: Boolean) = copy(useTestClock = useTestClock)
+    fun setInitialiseSerialization(initialiseSerialization: Boolean) = copy(initialiseSerialization = initialiseSerialization)
+    fun setNetworkMapStartStrategy(networkMapStartStrategy: NetworkMapStartStrategy) = copy(networkMapStartStrategy = networkMapStartStrategy)
+    fun setStartNodesInProcess(startNodesInProcess: Boolean) = copy(startNodesInProcess = startNodesInProcess)
+}
+
+private val defaultParameters = DriverParameters()
```
Comment (exFalso) [inline test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt]: I think the `defaultParameters` driver argument is better. If we have a static default then the situation with driverDirectory is unclear. Actually just realised that other parameters like portAllocation also don't work with the static solution!
If it's a parameter it also reduces the code required in the builder variant

**Comment by exFalso** [comment]
I'd approve but I can't!

**Pull request merged**